Add support for VLAN anti-spoof.

stephen-hurd opened the pull request:
This also enables VLAN insert and strip control to function correctly.

Commit:
Add support for VLAN anti-spoof.

This also enables VLAN insert and strip control to function correctly.

ajitkhaparde commented on drivers/net/bnxt/bnxt_ethdev.c:
```diff
@@ -1115,27 +1116,68 @@ static int bnxt_set_vf_vlan_filter_op(struct rte_eth_dev *dev, uint16_t vlan,
+					continue;
+
+				/* Now check that there's space */
+				if (bp->pf.vf_info[i].vlan_count == getpagesize() / sizeof(struct bnxt_vlan_table_entry)) {
```
So this is not based on hardware resource limit!

stephen-hurd commented:
Not currently... there doesn't seem to be a meaningful way to calculate
that.

Also, if a set of VLAN IDs can be reduced to a mask, a single rule is used
(or was at one point).

On Wed, Apr 26, 2017 at 2:47 PM, Ajit Khaparde <notifications@github.com>
wrote:

> *@ajitkhaparde* commented on this pull request.
> ------------------------------
>
> In drivers/net/bnxt/bnxt_ethdev.c
> <https://github.com/Broadcom/dpdk/pull/51#discussion_r113571021>:
>
> >  				continue;
> -			ret = bnxt_hwrm_set_vf_vlan(bp, i);
> -			if (ret)
> -				rc = ret;
> +			if (vlan_on) {
> +				/* First, search for a duplicate... */
> +				for (j=0; j<bp->pf.vf_info[i].vlan_count; j++) {
> +					if (bp->pf.vf_info[i].vlan_table[j].vid == vlan)
> +						continue;
> +				}
> +				if (j < bp->pf.vf_info[i].vlan_count)
> +					continue;
> +
> +				/* Now check that there's space */
> +				if (bp->pf.vf_info[i].vlan_count == getpagesize() / sizeof(struct bnxt_vlan_table_entry)) {
>
> So this is not based on hardware resource limit!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Broadcom/dpdk/pull/51#pullrequestreview-34971842>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ASsTfVdaWNAB7n8hhgWcyp1mKgpd48Buks5rz7t2gaJpZM4NJYRp>
> .
>



-- 
Stephen Hurd
949-926-8039
stephen.hurd@broadcom.com

ajitkhaparde commented on drivers/net/bnxt/bnxt_ethdev.c:
```diff
@@ -1115,27 +1116,68 @@ static int bnxt_set_vf_vlan_filter_op(struct rte_eth_dev *dev, uint16_t vlan,
+					memmove(&bp->pf.vf_info[i].vlan_table[j], &bp->pf.vf_info[i].vlan_table[j+1],
+					    getpagesize() - ((j+1) * sizeof(struct bnxt_vlan_table_entry)));
+					j--;
+				}
```
We need to decrement the vlan_count as well.

ajitkhaparde commented on drivers/net/bnxt/bnxt_ethdev.c:
```diff
@@ -1115,27 +1116,68 @@ static int bnxt_set_vf_vlan_filter_op(struct rte_eth_dev *dev, uint16_t vlan,
+			}
+			else {
+				if (bnxt_hwrm_vnic_qcfg(bp, &vnic, dflt_vnic) == 0) {
+					bnxt_hwrm_cfa_l2_set_rx_mask(bp, &vnic, bp->pf.vf_info[i].vlan_count, bp->pf.vf_info[i].vlan_table);
```
We are not returning the status in any of the above failures or if any of the calls fail.

stephen-hurd replied:
I'm really not sure what we could expect the caller to do with an error code though.  We can't roll back, and we can't communicate the new state.

I suppose the caller could exit...

Commit:
Decrement the VLAN count when one is removed

Commit:
Return -1 if the entire VLAN configuration can't be applied.

There is no way to recover or query the new state... but it's likely
that Everything is Fine.

ajitkhaparde commented:
Yes. That would be upto the application though.

On Wed, Apr 26, 2017 at 5:41 PM, Stephen Hurd <notifications@github.com>
wrote:

> *@stephen-hurd* commented on this pull request.
> ------------------------------
>
> In drivers/net/bnxt/bnxt_ethdev.c
> <https://github.com/Broadcom/dpdk/pull/51#discussion_r113579870>:
>
> > +			else {
> +				for (j=0; j<bp->pf.vf_info[i].vlan_count; j++) {
> +					if (bp->pf.vf_info[i].vlan_table[j].vid != vlan)
> +						continue;
> +					memmove(&bp->pf.vf_info[i].vlan_table[j], &bp->pf.vf_info[i].vlan_table[j+1],
> +					    getpagesize() - ((j+1) * sizeof(struct bnxt_vlan_table_entry)));
> +					j--;
> +				}
> +			}
> +			dflt_vnic = bnxt_hwrm_func_qcfg_vf_dflt_vnic_id(bp, i);
> +			if (dflt_vnic < 0) {
> +				RTE_LOG(ERR, PMD, "Unable to get default VNIC for VF %d\n", i);
> +			}
> +			else {
> +				if (bnxt_hwrm_vnic_qcfg(bp, &vnic, dflt_vnic) == 0) {
> +					bnxt_hwrm_cfa_l2_set_rx_mask(bp, &vnic, bp->pf.vf_info[i].vlan_count, bp->pf.vf_info[i].vlan_table);
>
> I'm really not sure what we could expect the caller to do with an error
> code though. We can't roll back, and we can't communicate the new state.
>
> I suppose the caller could exit...
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/Broadcom/dpdk/pull/51#discussion_r113579870>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AXqwrDzOH1FqFK0EowKj2rbXp_yYb0kPks5rz8gegaJpZM4NJYRp>
> .
>

Commit:
Merge branch '16.11_base' into vlan_antispoof_strip_insert
